src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs

```diff
@@ -30,7 +31,11 @@ internal abstract class ServiceConnectionBase : IServiceConnection
         private readonly ReadOnlyMemory<byte> _cachedPingBytes;
         private readonly HandshakeRequestMessage _handshakeRequest;
 
-        private readonly SemaphoreSlim _writeLock = new SemaphoreSlim(1, 1);
+        private readonly SemaphoreSlim _closeConnectionLock = new SemaphoreSlim(1, 1);
```

Comment:
Why do you need 4 locks? They seem unnecessary.

Reply:
`_closeConnectionLock` to ensure at most one `CloseAsync` process for each connection.
`_incomingLock` and `_outcoming` lock are used for synchronizing the `CloseAsync` process.
`_serverConnectionLock` is used for avoiding collisions during the `WriteAsync` method and it is already there.

Reply:
Turned the locks into awaiting Tasks after discussion.